libs/native-types/src/mssql.rs

```diff
@@ -0,0 +1,54 @@
+use serde::*;
+use serde_json::Value;
+use std::fmt;
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Copy)]
+pub enum DataLength {
+    Limited(u16),
+    Max,
+}
+
+impl fmt::Display for DataLength {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Limited(length) => write!(f, "{}", length),
+            Self::Max => write!(f, "max"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum MssqlType {
+    TinyInt,
+    SmallInt,
+    Int,
+    BigInt,
+    Decimal(u8, u8),
+    Numeric(u8, u8),
+    Money,
+    SmallMoney,
+    Bit,
+    Float(u8),
+    Real,
+    Date,
+    Time,
+    Datetime,
+    Datetime2,
+    DatetimeOffset,
+    SmallDatetime,
+    Char(DataLength),
+    VarChar(DataLength),
+    Text,
+    NVarChar(DataLength),
+    NText,
+    Binary(DataLength),
+    VarBinary(DataLength),
```

Comment:
Any reason not to express Max as a number and choose a number type that fits it? There is a bunch of native types code that relies on the length args being vec<u32>.

Reply:
But `MAX` needs to be rendered as `MAX`. It is a size in bytes, and `NVARCHAR` and `VARCHAR` fit different amount of text in the `MAX` variant.

Reply:
So ~2M characters in UTF-16 and ~4M characters in ASCII.

Reply:
Oh that is unfortunate. I was hoping to get away with just allowing integer arguments. I suggest to do the following:
* Keep this code as is because it's the right way to do it. It should reflect the data types of a given database as close as possible.
* In the parser we will only allow to supply integers in the beginning. We will then schedule work to support `@mydb.VarChar(MAX)` additionally.

Reply:
Captured in this issue: https://github.com/prisma/prisma-engines/issues/1145

Reply:
Some funny details about the Redmond database limits, facts time:

- `VARCHAR(n)` where `n` can be maximum of `4000`. This is the maximum number that fits into a row, is more performant and allows certain operations.
- `VARCHAR(max)` is then storing a pointer to the row that points to the heap that can hold two gigabytes of text per item.
- `NVARCHAR(n)` uses double the amount of space, so the maximum `n` is `2000`. The `max` variant works the same and gives two gigabytes of storage, that can of course store half the characters due to the data using more space.
- `VARBINARY(n)`, meaning we store bytes and the maximum `n` is `2000`. The `max` variant works here too.
- `TEXT` and `NTEXT` are a bit like `VARCHAR(max)` and `NVARCHAR(max)` but deprecated.
- `IMAGE` is a bit like `VARBINARY(max)` but deprecated.

You can use `(N)VARCHAR` as a primary key, but not the `max` variant.

Reply:
And extra funny fact for the fans:

With Microsoft SQL Server 2019, you can set a special collation for the database of `LATIN1_GENERAL_100_CI_AS_SC_UTF8`. If set, the `VARCHAR` fields accept UTF-8 data and we can completely forget the `N` variants. This is the reasonable thing to do and a very nice selling point to our native types, so you don't need to waste space with UCS-2.